File: python/voxgig_sekreto/plugins/gcpsecrets.py
# The gcpsecrets plugin: GCP Secret Manager. Needs HTTPS. A port of
# typescript/plugins/gcpsecrets.ts.
import base64
import os
import time

from ..addr import checkaddr
from ..sekreto import SekretoError, flatname
from ..providers import Provider, providerplugin
from .httpjson import fetchjson, tonumber


class GcpsecretsProvider(Provider):
    """GCP Secret Manager.

    `api.token` reads secret `api_token` (dots flattened to `_`; Secret
    Manager ids have no hierarchy and reject dots), latest version. The
    token comes from config, then `GOOGLE_OAUTH_ACCESS_TOKEN`, then the
    GCE/GKE metadata server - so on Google's own platform no credential
    configuration is needed at all.

    The metadata call itself is plain http to a link-local host by
    platform design; no credential rides on it, so `checkaddr` guards the
    Secret Manager address instead.
    """

    def __init__(self, options=None):
        self.opts = options or {}

        # A configured token is kept forever; a metadata-server token carries
        # expires_in and is renewed shortly before it runs out.
        self.livetoken = None
        self.renewat = float('inf')

    def metadataaddr(self):
        if self.opts.get('metadataaddr'):
            return self.opts['metadataaddr']
        host = os.environ.get('GCE_METADATA_HOST')
        return 'http://' + host if host else 'http://metadata.google.internal'

    def login(self):
        configured = self.opts.get('token') or os.environ.get('GOOGLE_OAUTH_ACCESS_TOKEN')
        if configured:
            return configured

        url = (self.metadataaddr().rstrip('/')
               + '/computeMetadata/v1/instance/service-accounts/default/token')

        status, body = fetchjson('GET', url, {'Metadata-Flavor': 'Google'})

        got = (body or {}).get('access_token')
        if 200 != status or not got:
            raise SekretoError('sekreto: gcp: no token and metadata server did not answer')

        expires = tonumber((body or {}).get('expires_in'))
        self.renewat = time.time() + max(expires - 60, 1) if 0 < expires else float('inf')

        return str(got)

    def lookup(self, name):
        project = self.opts.get('project') or ''
        if '' == project:
            raise SekretoError('sekreto: gcp: no project')

        addr = self.opts.get('addr') or 'https://secretmanager.googleapis.com'
        checkaddr(addr)

        if self.livetoken is None or time.time() >= self.renewat:
            self.livetoken = self.login()

        url = (addr.rstrip('/') + '/v1/projects/' + project + '/secrets/'
               + flatname(name, '_') + '/versions/latest:access')

        status, body = fetchjson('GET', url, {'authorization': 'Bearer ' + self.livetoken})

        if 404 == status:
            return None

        if 200 != status:
            raise SekretoError('sekreto: gcp error: ' + str(status) + ': ' + url)

        data = ((body or {}).get('payload') or {}).get('data')
        if not isinstance(data, str):
            return None

        # See the aws provider: validate=True, and an undecodable payload
        # is an error rather than a miss.
        try:
            return base64.b64decode(data, validate=True).decode('utf8')
        except Exception:
            raise SekretoError('sekreto: gcp: undecodable secret')

    def describe(self):
        return 'gcpsecrets:' + (self.opts.get('project') or '')


# The plugin: the `gcpsecrets` provider kind, as a voxgig/plugin definition.
gcpsecrets = providerplugin('gcpsecrets', GcpsecretsProvider)

File: ruby/lib/voxgig_sekreto/plugins/gcpsecrets.rb
# frozen_string_literal: true

# The gcpsecrets plugin: GCP Secret Manager. Needs HTTPS. A port of
# typescript/plugins/gcpsecrets.ts, which is canonical.

require_relative '../../voxgig_sekreto'
require_relative '../addr'
require_relative 'httpjson'

module VoxgigSekreto
  # GCP Secret Manager.
  #
  # `api.token` reads secret `api_token` (dots flattened to `_`; Secret
  # Manager ids have no hierarchy and reject dots), latest version. The
  # token comes from config, then `GOOGLE_OAUTH_ACCESS_TOKEN`, then the
  # GCE/GKE metadata server - so on Google's own platform no credential
  # configuration is needed at all.
  #
  # The metadata call itself is plain http to a link-local host by
  # platform design; no credential rides on it, so `checkaddr` guards the
  # Secret Manager address instead.
  class GcpsecretsProvider
    def initialize(opts = nil)
      @opts = opts || {}

      # A configured token is kept forever; a metadata-server token carries
      # expires_in and is renewed shortly before it runs out.
      @livetoken = nil
      @renewat = Float::INFINITY
    end

    def metadataaddr
      return @opts['metadataaddr'] if VoxgigSekreto.given(@opts['metadataaddr'])

      host = ENV.fetch('GCE_METADATA_HOST', nil)
      VoxgigSekreto.given(host) ? 'http://' + host : 'http://metadata.google.internal'
    end

    def login
      configured = @opts['token'] || ENV.fetch('GOOGLE_OAUTH_ACCESS_TOKEN', nil)
      return configured if VoxgigSekreto.given(configured)

      url = metadataaddr.sub(%r{/\z}, '') +
            '/computeMetadata/v1/instance/service-accounts/default/token'

      status, body = VoxgigSekreto.fetchjson('GET', url, { 'Metadata-Flavor' => 'Google' })

      got = body.is_a?(Hash) ? body['access_token'] : nil
      unless 200 == status && VoxgigSekreto.given(got)
        raise SekretoError, 'sekreto: gcp: no token and metadata server did not answer'
      end

      expires = body['expires_in'].to_f
      @renewat = 0 < expires ? Time.now.to_f + [expires - 60, 1].max : Float::INFINITY

      got.to_s
    end

    def lookup(name)
      project = @opts['project'] || ''
      raise SekretoError, 'sekreto: gcp: no project' if '' == project

      addr = @opts['addr'] || 'https://secretmanager.googleapis.com'
      VoxgigSekreto.checkaddr(addr)

      @livetoken = login if @livetoken.nil? || Time.now.to_f >= @renewat

      url = addr.sub(%r{/\z}, '') + '/v1/projects/' + project + '/secrets/' +
            VoxgigSekreto.flatname(name, '_') + '/versions/latest:access'

      status, body = VoxgigSekreto.fetchjson('GET', url,
                                             { 'authorization' => 'Bearer ' + @livetoken })

      return nil if 404 == status

      raise SekretoError, 'sekreto: gcp error: ' + status.to_s + ': ' + url if 200 != status

      data = body.is_a?(Hash) && body['payload'].is_a?(Hash) ? body['payload']['data'] : nil
      return nil unless data.is_a?(String)

      # See the aws provider: strict, and an undecodable payload is an
      # error rather than a miss.
      decoded = begin
        data.unpack1('m0')
      rescue ArgumentError
        raise SekretoError, 'sekreto: gcp: undecodable secret'
      end

      decoded.force_encoding('UTF-8')
    end

    def describe
      'gcpsecrets:' + (@opts['project'] || '')
    end
  end

  module Plugins
    # The plugin: the `gcpsecrets` provider kind, as a voxgig/plugin
    # definition.
    GCPSECRETS = VoxgigSekreto.providerplugin('gcpsecrets', lambda { |spec|
      GcpsecretsProvider.new(spec.slice('project', 'token', 'addr', 'metadataaddr'))
    })
  end
end
